tests/test_cse.py

```diff
@@ -0,0 +1,55 @@
+import numpy as np
+
+from examples.Acoustic_codegen import Acoustic_cg
+from examples.acoustic_example import source
+from examples.containers import IShot, IGrid
+
+
+def run_acoustic_forward(cse):
+    dimensions = (50, 50, 50)
+    model = IGrid()
+    model0 = IGrid()
+    model1 = IGrid()
+    model.shape = dimensions
+    model0.shape = dimensions
+    model1.shape = dimensions
+    origin = (0., 0.)
+    spacing = (20., 20.)
+
+    # True velocity
+    true_vp = np.ones(dimensions) + 2.0
+    true_vp[:, :, int(dimensions[0] / 2):int(dimensions[0])] = 4.5
+
+    model.create_model(origin, spacing, true_vp)
+
+    # Define seismic data.
+    data = IShot()
+
+    f0 = .010
+    dt = model.get_critical_dt()
+    t0 = 0.0
+    tn = 250.0
+    nt = int(1+(tn-t0)/dt)
+
+    time_series = source(np.linspace(t0, tn, nt), f0)
+    location = (origin[0] + dimensions[0] * spacing[0] * 0.5, 500,
+                origin[1] + 2 * spacing[1])
+    data.set_source(time_series, dt, location)
+
+    receiver_coords = np.zeros((101, 3))
+    receiver_coords[:, 0] = np.linspace(50, 950, num=101)
+    receiver_coords[:, 1] = 500
+    receiver_coords[:, 2] = location[2]
+    data.set_receiver_pos(receiver_coords)
+    data.set_shape(nt, 101)
+    Acoustic = Acoustic_cg(model, data)
+    (rec, u) = Acoustic.Forward(save=True, cse=cse)
+
```

Comment:
`save=True` is not necessary unless you want to compare the actual wavefield with
`assert np.isclose(np.linalg.norm(u- u_cse), 0.0)`
If not set to `False` as Travis really don't like large arrays